oauth2_provider/oauth2_validators.py

```diff
@@ -122,6 +123,11 @@ def _authenticate_basic_auth(self, request):
         elif request.client.client_id != client_id:
             log.debug("Failed basic auth: wrong client id %s" % client_id)
             return False
+        elif "$" in request.client.client_secret:
```

Comment:
that's definitely a risk. If the secret was generated using the default generator in this library, it should never contain a `$`. https://github.com/oauthlib/oauthlib/blob/master/oauthlib/common.py#L27

Reply:
I've included code comments to address this possibility. I don't see any other way to know whether a string has been hashed or not. I did add an additional equality check so that at least it's performing the same check that the next `elif` conditional does, so it should not be any "worse".